### PR TITLE
Support Elasticsearch Scroll API

### PR DESCRIPTION
**Pull request by fulmicoton**
Solution described in docs/internals/scroll.md.

# Tests

- Unit tests for the simplistic KV score.
- Unit test for the scroll API
- Rest API test comparing output from ElasticSearch.

### CODE REVIEW

File: quickwit/quickwit-common/src/duration.rs [hunk 1]
```diff
@@ -0,0 +1,71 @@
+// Copyright (C) 2023 Quickwit, Inc.
```
Comment (guilload) [inline quickwit/quickwit-common/src/duration.rs:1]: We have a dependency on [`humantime`](https://crates.io/crates/humantime) now, I don't think we need this anymore. I pinged the owner of the crate about the maintenance status of the project.

Comment (guilload) [inline quickwit/quickwit-common/src/duration.rs:1]: `human_duration.rs`

File: quickwit/quickwit-proto/protos/quickwit/search_api.proto [hunk 2]
```diff
@@ -110,6 +132,11 @@ message SearchRequest {
 
   // Fields to extract snippet on
   repeated string  snippet_fields = 12;
+
+  // If set, the search response will include a search id
+  // that will make it possible to paginate through the results
+  // in a consistent manner.
+  optional uint32 scroll_secs = 14;
```
Comment (guilload) [inline quickwit/quickwit-proto/protos/quickwit/search_api.proto:139]:
```suggestion
  optional uint32 scroll_ttl_secs = 14;
```

File: quickwit/quickwit-search/Cargo.toml [hunk 3]
```diff
@@ -33,9 +33,11 @@ tokio = { workspace = true }
 tokio-stream = { workspace = true }
 tokio-util = { workspace = true }
 tower = { workspace = true }
+ttl_cache = { workspace = true }
```
Comment (guilload) [inline quickwit/quickwit-search/Cargo.toml:36]: Last released 4 years ago and is based on an also unmaintained crate. Let's keep that in mind.

File: quickwit/quickwit-search/src/cluster_client.rs [hunk 4]
```diff
@@ -157,6 +161,35 @@ impl ClusterClient {
...
+    /// This function may fail silently, if no clients was available.
+    pub async fn put_search_after_context(
+        &self,
+        put_search_after_context_req: PutSearchAfterContextRequest,
+    ) {
+        let clients = self
+            .search_job_placer
+            .best_nodes_per_affinity(&put_search_after_context_req.search_after_id)
+            .await;
+        let mut successful_replication = 0;
```
Comment (guilload) [inline quickwit/quickwit-search/src/cluster_client.rs:176]:
```suggestion
        // TODO: Issue concurrent requests
```

File: quickwit/quickwit-search/src/search_after_cache.rs [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        SearchAfterCache {
+            ttl_with_cache: Arc::new(RwLock::new(TtlCache::new(SEARCH_AFTER_CONTEXT_CAPACITY))),
+        }
+    }
+}
+
+impl SearchAfterCache {
+    pub async fn put(&self, key: String, payload: &SearchAfterContext, ttl: Duration) {
+        // TODO add light compression
+        let mut cache_lock = self.ttl_with_cache.write().await;
```
Comment (guilload) [inline quickwit/quickwit-search/src/search_after_cache.rs:53]: You should encode first and then grab the lock.

File: quickwit/quickwit-serve/src/elastic_search_api/rest_handler.rs [hunk 6]
```diff
@@ -135,6 +135,9 @@ fn build_request_for_es_api(
         (None, None)
     };
 
+    let scroll_duration: Option<Duration> = search_params.get_scroll_duration()?;
```
Comment (guilload) [inline quickwit/quickwit-serve/src/elastic_search_api/rest_handler.rs:138]:
```suggestion
    let scroll_ttl: Option<Duration> = search_params.parse_scroll_ttl()?;
```

File: quickwit/quickwit-serve/src/elastic_search_api/rest_handler.rs [hunk 7]
```diff
@@ -135,6 +135,9 @@ fn build_request_for_es_api(
...
+    let scroll_secs: Option<u32> = scroll_duration.map(|duration| duration.as_secs() as u32);
```
Comment (guilload) [inline quickwit/quickwit-serve/src/elastic_search_api/rest_handler.rs:139]:
```suggestion
    let scroll_ttl_secs: Option<u32> = scroll_ttl.map(|duration| duration.as_secs() as u32);
```

File: quickwit/quickwit-serve/src/elastic_search_api/model/search_query_params.rs [hunk 8]
```diff
@@ -189,6 +191,18 @@ impl SearchQueryParams {
         }
         Ok(Some(sort_fields))
     }
+
+    /// Returns the scroll duration supplied by the user.
+    ///
+    /// This function returns an error if the scroll duration is not in the expected format. (`40s`
+    /// etc.)
+    pub fn get_scroll_duration(&self) -> Result<Option<Duration>, SearchError> {
```
Comment (guilload) [inline quickwit/quickwit-serve/src/elastic_search_api/model/search_query_params.rs:199]:
```suggestion
    pub fn parse_scroll_ttl(&self) -> Result<Option<Duration>, SearchError> {
```

File: quickwit/quickwit-common/src/duration.rs (hunk 1 above)
Comment (fulmicoton) [reply quickwit/quickwit-common/src/duration.rs]: moved toward humantime.

File: quickwit/quickwit-proto/protos/quickwit/search_api.proto (hunk 2 above)
Comment (fulmicoton) [reply quickwit/quickwit-proto/protos/quickwit/search_api.proto]: renamed

File: quickwit/quickwit-serve/src/elastic_search_api/rest_handler.rs (hunk 6 above)
Comment (fulmicoton) [reply quickwit/quickwit-serve/src/elastic_search_api/rest_handler.rs:138]: obsolete.

File: quickwit/quickwit-search/src/cluster_client.rs (hunk 4 above)
Comment (guilload) [reply quickwit/quickwit-search/src/cluster_client.rs:173]: We don't have to issue the requests sequentially.

Comment (fulmicoton) [reply quickwit/quickwit-search/src/cluster_client.rs:173]: ah ok got it!

**Comment by fulmicoton** [comment]
@trinity-1686a can you review?

It is a weird functionality and a weird PR for it. The details are available in `scroll.md`.

**Comment by fulmicoton** [comment]
Closing for #3705

**Pull request closed without merging**